src/mpid/common/thread/mpidu_thread_fallback.h

```diff
@@ -252,24 +281,17 @@ M*/
 #define MPIDU_Thread_join       MPL_thread_join
 #define MPIDU_Thread_same       MPL_thread_same
 
-/*@
-  MPIDU_Thread_yield - voluntarily relinquish the CPU, giving other threads an opportunity to run
-@*/
-#define MPIDU_Thread_yield(mutex_ptr_, err_ptr_)                        \
-    do {                                                                \
-        int saved_count_ = (mutex_ptr_)->count;                         \
-        MPL_thread_id_t saved_owner_ = (mutex_ptr_)->owner;             \
-        MPIR_Assert(saved_count_ > 0);                                  \
-        (mutex_ptr_)->count = 0;                                        \
-        (mutex_ptr_)->owner = 0;                                        \
-        MPIDU_Thread_mutex_unlock(mutex_ptr_, err_ptr_);                \
-        MPIR_Assert(*err_ptr_ == 0);                                    \
-        MPL_thread_yield();                                             \
-        MPIDU_Thread_mutex_lock(mutex_ptr_, err_ptr_, MPL_THREAD_PRIO_LOW);\
-        MPIR_Assert((mutex_ptr_)->count == 0);                          \
-        (mutex_ptr_)->count = saved_count_;                             \
-        (mutex_ptr_)->owner = saved_owner_;                             \
-        MPIR_Assert(*err_ptr_ == 0);                                    \
+#define MPIDU_Thread_yield() \
+    do { \
+        if (MPIR_CVAR_ENABLE_HEAVY_YIELD) { \
+            /* note: sleep time may be rounded up to the granularity of the underlying clock */ \
+            struct timespec t; \
+            t.tv_sec = 0; \
+            t.tv_nsec = 1; \
+            nanosleep(&t, NULL); \
```

Comment:
we may also want to note that this yield method may be incompatible with non-pthread configurations.

Reply:
Updated the cvar description:
```
      description : >-
        If enabled, use nanosleep to ensure other threads have a chance to grab the lock.
        Note: this may not work with some thread runtimes, e.g. non-preemptive user-level
        threads.
```